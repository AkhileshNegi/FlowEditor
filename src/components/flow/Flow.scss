@import '~/variables.scss';

.loading {
    visibility: hidden;
    height: inherit;
}

.loaded {
    height: inherit;
    position: relative;
}

.nodeList {
    position: relative;
    user-select: none;
}

.dragSelection {
    position: absolute;
    background: red;
    border: 0px solid darken($selection_square, 10%);
    opacity: 0;
    pointer-events: none;
    z-index: $z_selection;
}
