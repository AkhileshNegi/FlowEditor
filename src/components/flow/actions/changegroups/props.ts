import { ActionFormHelper, Type } from '~/config/typeConfigs';
import { ChangeGroups } from '~/flowTypes';
import { NodeEditorSettings } from '~/store/nodeEditor';

export interface ChangeGroupsFormProps {
    // action details
    nodeSettings: NodeEditorSettings;
    formHelper: ActionFormHelper;
    typeConfig: Type;

    // update handlers
    updateAction(action: ChangeGroups): void;

    // modal notifiers
    onTypeChange(config: Type): void;
    onClose(canceled: boolean): void;
}
export default ChangeGroupsFormProps;
