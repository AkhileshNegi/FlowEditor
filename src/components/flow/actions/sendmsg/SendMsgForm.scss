@import '~/variables.scss';

.checkbox {
    margin-top: 20px;
}

.quick-reply-summary {
    margin-top: -20px;
    position: relative;
}

.type-choice {
    display: inline-block;
    width: 180px;
}

.url-attachment {
    margin-bottom: 8px;

    .upload {
        padding-left: 5px;
        vertical-align: super;
    }

    .remove-upload {
        display: inline-block;
    }

    .url {
        display: inline-block;
        margin-left: 8px;
        width: 329px;
        margin-right: 10px;

        input {
            vertical-align: super;
            height: 28px !important;
            margin-top: 3px;
            width: 100%;
            font-size: 12px;
        }
    }

    .remove {
        vertical-align: super;
        display: inline-block;
    }
}

.existing-quick-replies {
    padding: 5px;
    border-radius: 5px;
    border: 1px solid $light_gray;

    .existing-quick-reply {
        margin: 2px;
        display: inline-block;
    }
}

.add-quick-replies {
    margin-top: 10px;
}
