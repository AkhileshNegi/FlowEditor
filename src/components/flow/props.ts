import { UpdateLocalizations } from '~/components/nodeeditor/NodeEditor';
import { Type } from '~/config/typeConfigs';
import { AnyAction } from '~/flowTypes';
import { Asset } from '~/services/AssetService';
import { AssetStore, RenderNode } from '~/store/flowContext';
import { NodeEditorSettings } from '~/store/nodeEditor';
import { DispatchWithState, GetState } from '~/store/thunks';

export interface ActionFormProps {
    // action details
    nodeSettings: NodeEditorSettings;
    typeConfig: Type;
    assets: AssetStore;

    // update handlers
    updateAction(
        action: AnyAction,
        onUpdated?: (dispatch: DispatchWithState, getState: GetState) => void
    ): void;

    // modal notifiers
    onTypeChange(config: Type): void;
    onClose(canceled: boolean): void;
}

export interface RouterFormProps {
    nodeSettings: NodeEditorSettings;
    typeConfig: Type;

    assets: AssetStore;

    // update handlers
    updateRouter(renderNode: RenderNode): void;

    // modal notifiers
    onTypeChange(config: Type): void;
    onClose(canceled: boolean): void;
}

export interface LocalizationFormProps {
    language: Asset;
    nodeSettings: NodeEditorSettings;
    updateLocalizations(languageCode: string, localizations: any[]): UpdateLocalizations;
    onClose(canceled: boolean): void;
}
