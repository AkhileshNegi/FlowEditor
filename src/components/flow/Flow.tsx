import { react as bindCallbacks } from 'auto-bind';
import * as React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import * as styles from '~/components/flow/Flow.scss';
import ConnectedNode from '~/components/flow/node/Node';
import { getDraggedFrom } from '~/components/helpers';
import ConnectedNodeEditor from '~/components/nodeeditor/NodeEditor';
import Simulator from '~/components/simulator/Simulator';
import Sticky from '~/components/sticky/Sticky';
import { ConfigProviderContext } from '~/config';
import { fakePropType } from '~/config/ConfigProvider';
import { getActivity } from '~/external';
import { FlowDefinition, FlowPosition } from '~/flowTypes';
import ActivityManager from '~/services/ActivityManager';
import Plumber from '~/services/Plumber';
import { DragSelection, EditorState } from '~/store/editor';
import { RenderNode } from '~/store/flowContext';
import { getCollisions } from '~/store/helpers';
import { NodeEditorSettings } from '~/store/nodeEditor';
import AppState from '~/store/state';
import {
    ConnectionEvent,
    DispatchWithState,
    ensureStartNode,
    EnsureStartNode,
    mergeEditorState,
    MergeEditorState,
    NoParamsAC,
    OnConnectionDrag,
    onConnectionDrag,
    OnDragSelection,
    onDragSelection,
    OnOpenNodeEditor,
    onOpenNodeEditor,
    resetNodeEditingState,
    UpdateConnection,
    updateConnection,
    updateSticky,
    UpdateSticky
} from '~/store/thunks';
import { contextTypes } from '~/testUtils';
import {
    createUUID,
    isRealValue,
    NODE_PADDING,
    renderIf,
    snapToGrid,
    timeEnd,
    timeStart
} from '~/utils';
import Debug from '~/utils/debug';

export interface FlowStoreProps {
    editorState: Partial<EditorState>;
    mergeEditorState: MergeEditorState;

    definition: FlowDefinition;
    nodes: { [uuid: string]: RenderNode };
    dependencies: FlowDefinition[];
    nodeEditorSettings: NodeEditorSettings;

    ensureStartNode: EnsureStartNode;
    updateConnection: UpdateConnection;
    onOpenNodeEditor: OnOpenNodeEditor;
    onDragSelection: OnDragSelection;
    resetNodeEditingState: NoParamsAC;
    onConnectionDrag: OnConnectionDrag;
    updateSticky: UpdateSticky;
}

export interface Translations {
    [uuid: string]: any;
}

export const DRAG_THRESHOLD = 8;
export const REPAINT_TIMEOUT = 500;
export const GHOST_POSITION_INITIAL = { left: -1000, top: -1000 };

export const nodeSpecId = 'node';
export const nodesContainerSpecId = 'node-container';
export const ghostNodeSpecId = 'ghost-node';
export const dragSelectSpecId = 'drag-select';

export const isDraggingBack = (event: ConnectionEvent) => {
    return event.suspendedElementId === event.targetId && event.source !== null;
};

export const getDragStyle = (drag: DragSelection) => {
    const left = Math.min(drag.startX, drag.currentX);
    const top = Math.min(drag.startY, drag.currentY);
    const width = Math.max(drag.startX, drag.currentX) - left;
    const height = Math.max(drag.startY, drag.currentY) - top;
    return {
        left,
        top,
        width,
        height
    };
};

export class Flow extends React.Component<FlowStoreProps, {}> {
    private Activity: ActivityManager;
    private Plumber: Plumber;
    private containerOffset = { left: 0, top: 0 };
    private ele: HTMLDivElement;
    private nodeContainerUUID: string;

    // Refs
    private ghost: any;

    public static contextTypes = {
        endpoints: fakePropType,
        debug: fakePropType
    };

    constructor(props: FlowStoreProps, context: ConfigProviderContext) {
        super(props, context);

        this.nodeContainerUUID = createUUID();

        this.Activity = new ActivityManager(this.props.definition.uuid, getActivity);

        this.Plumber = new Plumber();

        /* istanbul ignore next */
        if (context.debug) {
            window.fe = new Debug(props, this.props.editorState.debug);
        }

        bindCallbacks(this, {
            include: [/Ref$/, /^on/, /^is/]
        });

        timeStart('RenderAndPlumb');
    }

    private onRef(ref: HTMLDivElement): HTMLDivElement {
        return (this.ele = ref);
    }

    private ghostRef(ref: any): any {
        return (this.ghost = ref);
    }

    public componentDidMount(): void {
        this.Plumber.bind('connection', (event: ConnectionEvent) =>
            this.props.updateConnection(event.sourceId, event.targetId)
        );
        this.Plumber.bind('beforeDrag', (event: ConnectionEvent) =>
            this.beforeConnectionDrag(event)
        );

        this.Plumber.bind('connectionDrag', (event: ConnectionEvent) => {
            this.props.onConnectionDrag(event);
        });

        this.Plumber.bind('connectionDragStop', (event: ConnectionEvent) =>
            this.onConnectorDrop(event)
        );
        this.Plumber.bind(
            'beforeStartDetach',
            (event: ConnectionEvent) => !this.props.editorState.translating
        );
        this.Plumber.bind('beforeDetach', (event: ConnectionEvent) => true);
        this.Plumber.bind('beforeDrop', (event: ConnectionEvent) =>
            this.onBeforeConnectorDrop(event)
        );

        // If we don't have any nodes, create our first one
        this.props.ensureStartNode();

        let offset = { left: 0, top: 0 };

        /* istanbul ignore next */
        if (this.ele) {
            offset = this.ele.getBoundingClientRect();
        }
        this.containerOffset = { left: offset.left, top: offset.top + window.scrollY };

        timeEnd('RenderAndPlumb');

        // deals with safari load rendering throwing
        // off the jsplumb offsets
        window.setTimeout(() => this.Plumber.repaint(), REPAINT_TIMEOUT);
    }

    public componentWillUnmount(): void {
        this.Plumber.reset();
    }

    /**
     * Called right before a connector is dropped onto a new node
     */
    private onBeforeConnectorDrop(event: ConnectionEvent): boolean {
        this.props.resetNodeEditingState();

        if (event.sourceId.split(':')[0] === event.targetId) {
            console.error(event.targetId + ' cannot point to itself');
            return false;
        }

        return true;
    }

    // private onShowDefinition(definition: FlowDefinition): void {
    //     TODO: make this work, it's cool!
    //     this.Plumber.reset();
    //     this.setState({ viewDefinition: definition }, () => { this.Plumber.repaint() });
    // }

    /**
     * Called the moment a connector is done dragging, whether it is dropped on an
     * existing node or on to empty space.
     */
    private onConnectorDrop(event: ConnectionEvent): boolean {
        const { ghostNode } = this.props.editorState;
        // Don't show the node editor if we a dragging back to where we were
        if (isRealValue(ghostNode) && !isDraggingBack(event)) {
            // Wire up the drag from to our ghost node
            this.Plumber.recalculate(ghostNode.node.uuid);

            const dragPoint = getDraggedFrom(ghostNode);
            this.Plumber.connect(
                dragPoint.nodeUUID + ':' + dragPoint.exitUUID,
                ghostNode.node.uuid
            );

            // Save our position for later
            const { left, top } = (this.ghost &&
                snapToGrid(
                    this.ghost.wrappedInstance.ele.offsetLeft,
                    this.ghost.wrappedInstance.ele.offsetTop
                )) || { left: 0, top: 0 };

            this.props.editorState.ghostNode.ui.position = { left, top };
            let originalAction = null;
            if (ghostNode.node.actions && ghostNode.node.actions.length === 1) {
                originalAction = ghostNode.node.actions[0];
            }

            // Bring up the node editor
            this.props.onOpenNodeEditor({
                originalNode: ghostNode,
                originalAction
            });
        }

        // To-do: mock this out
        /* istanbul ignore next */
        $(document).off('mousemove');

        return true;
    }

    private beforeConnectionDrag(event: ConnectionEvent): boolean {
        return !this.props.editorState.translating;
    }

    private getNodes(): JSX.Element[] {
        return Object.keys(this.props.nodes).map(uuid => {
            const renderNode = this.props.nodes[uuid];
            return (
                <div
                    key={uuid}
                    /*{...createClickHandler(() => {
                        console.log('do it');
                    })}*/

                    onMouseDown={(event: React.MouseEvent<HTMLDivElement>) => {
                        // this.props.mergeEditorState({});
                        this.onDragNodeStart(renderNode, {
                            left:
                                event.pageX -
                                this.containerOffset.left -
                                renderNode.ui.position.left,
                            top: event.pageY - this.containerOffset.top - renderNode.ui.position.top
                        });
                    }}
                    onMouseUp={(event: React.MouseEvent<HTMLDivElement>) => {
                        this.onDragNodeStop(renderNode.node.uuid);
                    }}
                >
                    <ConnectedNode
                        key={renderNode.node.uuid}
                        data-spec={nodeSpecId}
                        nodeUUID={renderNode.node.uuid}
                        Activity={this.Activity}
                        plumberRepaintForDuration={this.Plumber.repaintForDuration}
                        plumberMakeTarget={this.Plumber.makeTarget}
                        plumberRemove={this.Plumber.remove}
                        plumberRecalculate={this.Plumber.recalculate}
                        plumberMakeSource={this.Plumber.makeSource}
                        plumberConnectExit={this.Plumber.connectExit}
                        plumberUpdateClass={this.Plumber.updateClass}
                    />
                </div>
            );
        });
    }

    private getStickies(): JSX.Element[] {
        const stickyMap = this.props.definition._ui.stickies || {};
        return Object.keys(stickyMap).map(uuid => {
            return <Sticky key={uuid} uuid={uuid} sticky={stickyMap[uuid]} />;
        });
    }

    private getDragNode(): JSX.Element {
        return isRealValue(this.props.editorState.ghostNode) ? (
            <ConnectedNode
                ref={this.ghostRef}
                key={this.props.editorState.ghostNode.node.uuid}
                data-spec={ghostNodeSpecId}
                ghost={true}
                nodeUUID={this.props.editorState.ghostNode.node.uuid}
                Activity={this.Activity}
                plumberRepaintForDuration={this.Plumber.repaintForDuration}
                plumberMakeTarget={this.Plumber.makeTarget}
                plumberRemove={this.Plumber.remove}
                plumberRecalculate={this.Plumber.recalculate}
                plumberMakeSource={this.Plumber.makeSource}
                plumberConnectExit={this.Plumber.connectExit}
                plumberUpdateClass={this.Plumber.updateClass}
            />
        ) : null;
    }

    private getSimulator(): JSX.Element {
        return renderIf(this.context.endpoints && this.context.endpoints.simulateStart)(
            <Simulator Activity={this.Activity} />
        );
    }

    private getNodeEditor(): JSX.Element {
        return renderIf(this.props.nodeEditorSettings !== null)(
            <ConnectedNodeEditor
                plumberConnectExit={this.Plumber.connectExit}
                plumberRepaintForDuration={this.Plumber.repaintForDuration}
            />
        );
    }

    private isClickOnCanvas(event: React.MouseEvent<HTMLDivElement>): boolean {
        // TODO: not sure the TS-safe way to access id here
        return (event.target as any).id === this.nodeContainerUUID;
    }

    private onDoubleClick(event: React.MouseEvent<HTMLDivElement>): void {
        if (this.isClickOnCanvas(event)) {
            const { left, top } = snapToGrid(
                event.clientX - this.containerOffset.left - 100 + NODE_PADDING,
                event.clientY - this.containerOffset.top - NODE_PADDING * 2
            );

            this.props.updateSticky(createUUID(), {
                position: { left, top },
                title: 'New Note',
                body: '...'
            });
        }
    }

    public onMouseDown(event: React.MouseEvent<HTMLDivElement>): void {
        if (this.isClickOnCanvas(event)) {
            this.props.mergeEditorState({
                dragSelection: {
                    startX: event.pageX - this.containerOffset.left,
                    startY: event.pageY - this.containerOffset.top,
                    currentX: event.pageX - this.containerOffset.left,
                    currentY: event.pageY - this.containerOffset.top,
                    selected: null
                }
            });
        }
    }

    public onMouseMove(event: React.MouseEvent<HTMLDivElement>): void {
        if (this.props.editorState.dragSelection && this.props.editorState.dragSelection.startX) {
            const drag = this.props.editorState.dragSelection;
            const left = Math.min(drag.startX, drag.currentX);
            const top = Math.min(drag.startY, drag.currentY);
            const right = Math.max(drag.startX, drag.currentX);
            const bottom = Math.max(drag.startY, drag.currentY);

            this.props.mergeEditorState({
                dragSelection: {
                    startX: drag.startX,
                    startY: drag.startY,
                    currentX: event.pageX - this.containerOffset.left,
                    currentY: event.pageY - this.containerOffset.top,
                    selected: getCollisions(this.props.nodes, { left, top, right, bottom })
                }
            });
        }

        if (this.props.editorState.dragNodeUUID) {
            const startPosition = this.props.editorState.dragActive
                ? this.props.editorState.dragSelection.selected[this.props.editorState.dragNodeUUID]
                : this.props.nodes[this.props.editorState.dragNodeUUID].ui.position;

            const xd =
                event.pageX -
                this.containerOffset.left -
                this.props.editorState.dragDownPosition.left -
                startPosition.left;

            const yd =
                event.pageY -
                this.containerOffset.top -
                this.props.editorState.dragDownPosition.top -
                startPosition.top;

            if (this.props.editorState.dragActive) {
                this.props.onDragSelection({ left: xd, top: yd });
            } else {
                if (Math.abs(xd) + Math.abs(yd) > DRAG_THRESHOLD) {
                    const renderNode = this.props.nodes[this.props.editorState.dragNodeUUID];
                    const selection = this.props.editorState.dragSelection || {};
                    let selected = selection.selected || {};
                    if (!(renderNode.node.uuid in selected)) {
                        selected = { [renderNode.node.uuid]: renderNode.ui.position };
                    }

                    this.props.mergeEditorState({
                        dragActive: true,
                        dragSelection: {
                            selected
                        }
                    });
                }
            }
        }
    }

    public onMouseUp(event: React.MouseEvent<HTMLDivElement>): void {
        if (this.props.editorState.dragNodeUUID) {
            this.props.mergeEditorState({
                dragNodeUUID: null
            });
        }

        if (this.props.editorState.dragSelection && this.props.editorState.dragSelection.startX) {
            this.props.mergeEditorState({
                dragSelection: {
                    startX: null,
                    startY: null,
                    currentX: null,
                    currentY: null,
                    selected: this.props.editorState.dragSelection.selected
                }
            });

            // if (this.props.editorState.dragSelection.selected) {
            // this.Plumber.setDragSelection(this.props.editorState.dragSelection.selected);
            // }
        }
    }

    public getDragSelectionBox(): JSX.Element {
        if (this.props.editorState.dragSelection && this.props.editorState.dragSelection.startX) {
            return (
                <div
                    data-spec={dragSelectSpecId}
                    className={styles.dragSelection}
                    style={{ ...getDragStyle(this.props.editorState.dragSelection) }}
                />
            );
        }
        return null;
    }

    private onDragNodeStart(renderNode: RenderNode, position: FlowPosition): void {
        this.props.mergeEditorState({
            dragNodeUUID: renderNode.node.uuid,
            dragDownPosition: position
        });
    }

    private onDragNodeDrag(e: React.MouseEvent<HTMLDivElement>): void {}

    private onDragNodeStop(uuid: string): void {
        console.log('stopping:', uuid);
        // this.setState({ thisNodeDragging: false });
        this.props.mergeEditorState({
            dragNodeUUID: null,
            dragDownPosition: null,
            dragSelection: null,
            dragActive: false
        });
    }

    public render(): JSX.Element {
        return (
            <>
                {this.getSimulator()}
                {this.getDragNode()}
                {this.getNodeEditor()}
                <div
                    ref={this.onRef}
                    id={this.nodeContainerUUID}
                    className={styles.nodeList}
                    data-spec={nodesContainerSpecId}
                    onMouseDown={this.onMouseDown}
                    onMouseMove={this.onMouseMove}
                    onMouseUp={this.onMouseUp}
                    onDoubleClick={this.onDoubleClick}
                >
                    {this.getStickies()}
                    {this.getDragSelectionBox()}
                    {this.getNodes()}
                </div>
            </>
        );
    }
}

/* istanbul ignore next */
const mapStateToProps = ({
    flowContext: { definition, dependencies, nodes },
    editorState,
    nodeEditor: { settings }
}: AppState) => {
    return {
        nodeEditorSettings: settings,
        definition,
        nodes,
        dependencies,
        editorState: editorState as Partial<EditorState>
    };
};

/* istanbul ignore next */
const mapDispatchToProps = (dispatch: DispatchWithState) =>
    bindActionCreators(
        {
            mergeEditorState,
            ensureStartNode,
            resetNodeEditingState,
            onConnectionDrag,
            onOpenNodeEditor,
            onDragSelection,
            updateConnection,
            updateSticky
        },
        dispatch
    );

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(Flow);
