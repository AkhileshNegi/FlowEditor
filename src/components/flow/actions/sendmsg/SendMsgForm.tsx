import { react as bindCallbacks } from 'auto-bind';
import axios from 'axios';
import mutate from 'immutability-helper';
import * as React from 'react';
import Dialog, { ButtonSet, Tab } from '~/components/dialog/Dialog';
import { initializeForm, stateToAction } from '~/components/flow/actions/sendmsg/helpers';
import * as localStyles from '~/components/flow/actions/sendmsg/SendMsgForm.scss';
import { ActionFormProps } from '~/components/flow/props';
import CheckboxElement from '~/components/form/checkbox/CheckboxElement';
import SelectElement, { SelectOption } from '~/components/form/select/SelectElement';
import TaggingElement from '~/components/form/select/tags/TaggingElement';
import TextInputElement, { Count } from '~/components/form/textinput/TextInputElement';
import TypeList from '~/components/nodeeditor/TypeList';
import Pill from '~/components/pill/Pill';
import { fakePropType } from '~/config/ConfigProvider';
import { FormState, mergeForm, StringArrayEntry, StringEntry } from '~/store/nodeEditor';
import { validate, validateMaxOfTen, validateRequired } from '~/store/validators';
import { createUUID } from '~/utils';
import { small } from '~/utils/reactselect';

import * as styles from './SendMsgForm.scss';

const MAX_ATTACHMENTS = 5;

const TYPE_OPTIONS: SelectOption[] = [
    { value: 'image', label: 'Image URL' },
    { value: 'audio', label: 'Audio URL' },
    { value: 'video', label: 'Video URL' }
];

const NEW_TYPE_OPTIONS = TYPE_OPTIONS.concat([{ value: 'upload', label: 'Upload Attachment' }]);

const getAttachmentTypeOption = (type: string): SelectOption => {
    return TYPE_OPTIONS.find((option: SelectOption) => option.value === type);
};

export interface Attachment {
    type: string;
    url: string;
    uploaded?: boolean;
}

export interface SendMsgFormState extends FormState {
    text: StringEntry;
    quickReplies: StringArrayEntry;
    sendAll: boolean;
    attachments: Attachment[];
}

export default class SendMsgForm extends React.Component<ActionFormProps, SendMsgFormState> {
    private filePicker: any;

    constructor(props: ActionFormProps) {
        super(props);
        this.state = initializeForm(this.props.nodeSettings);
        bindCallbacks(this, {
            include: [/^handle/, /^on/]
        });
    }

    public static contextTypes = {
        endpoints: fakePropType
    };

    private handleUpdate(keys: {
        text?: string;
        sendAll?: boolean;
        quickReplies?: string[];
    }): boolean {
        const updates: Partial<SendMsgFormState> = {};

        if (keys.hasOwnProperty('text')) {
            updates.text = validate('Message', keys.text, [validateRequired]);
        }

        if (keys.hasOwnProperty('sendAll')) {
            updates.sendAll = keys.sendAll;
        }

        if (keys.hasOwnProperty('quickReplies')) {
            updates.quickReplies = validate('Quick Replies', keys.quickReplies, [validateMaxOfTen]);
        }

        const updated = mergeForm(this.state, updates);
        this.setState(updated);
        return updated.valid;
    }

    public handleMessageUpdate(text: string): boolean {
        return this.handleUpdate({ text });
    }

    public handleQuickRepliesUpdate(quickReplies: string[]): boolean {
        return this.handleUpdate({ quickReplies });
    }

    public handleSendAllUpdate(sendAll: boolean): boolean {
        return this.handleUpdate({ sendAll });
    }

    private handleSave(): void {
        // make sure we validate untouched text fields
        const valid = this.handleUpdate({
            text: this.state.text.value
        });

        if (valid) {
            this.props.updateAction(stateToAction(this.props.nodeSettings, this.state));

            // notify our modal we are done
            this.props.onClose(false);
        }
    }

    public handleAttachmentRemoved(index: number): void {
        // we found a match, merge us in
        const updated: any = mutate(this.state.attachments, {
            $splice: [[index, 1]]
        });
        this.setState({ attachments: updated });
    }

    private getButtons(): ButtonSet {
        return {
            primary: { name: 'Ok', onClick: this.handleSave },
            secondary: { name: 'Cancel', onClick: () => this.props.onClose(true) }
        };
    }

    private renderUpload(index: number, attachment: Attachment): JSX.Element {
        return (
            <div
                className={styles.urlAttachment}
                key={index > -1 ? 'url_attachment_' + index : createUUID()}
            >
                <div className={styles.typeChoice}>
                    <SelectElement
                        name="Type"
                        styles={small}
                        entry={{
                            value: { label: attachment.type }
                        }}
                        options={TYPE_OPTIONS}
                    />
                </div>
                <div className={styles.url}>
                    <span className={styles.upload}>
                        <Pill
                            icon="fe-download"
                            text=" Download"
                            large={true}
                            onClick={() => {
                                window.open(attachment.url, '_blank');
                            }}
                        />
                        <div className={styles.removeUpload}>
                            <Pill
                                icon="fe-x"
                                text=" Remove"
                                large={true}
                                onClick={() => {
                                    this.handleAttachmentRemoved(index);
                                }}
                            />
                        </div>
                    </span>
                </div>
            </div>
        );
    }

    private handleUploadFile(files: FileList): void {
        let attachments: any = this.state.attachments;

        const data = new FormData();
        data.append('file', files[0]);
        axios
            .post(this.context.endpoints.attachments, data)
            .then(response => {
                attachments = mutate(attachments, {
                    $push: [{ type: response.data.type, url: response.data.url, uploaded: true }]
                });
                this.setState({ attachments });
            })
            .catch(error => {
                console.log(error);
            });
    }

    private renderAttachment(index: number, attachment: Attachment): JSX.Element {
        let attachments: any = this.state.attachments;
        return (
            <div
                className={styles.urlAttachment}
                key={index > -1 ? 'url_attachment_' + index : createUUID()}
            >
                <div className={styles.typeChoice}>
                    <SelectElement
                        styles={small}
                        name="Type Options"
                        placeholder="Add Attachment"
                        entry={{
                            value: index > -1 ? getAttachmentTypeOption(attachment.type) : null
                        }}
                        onChange={(option: any) => {
                            if (option.value === 'upload') {
                                window.setTimeout(() => {
                                    this.filePicker.click();
                                }, 200);
                            } else {
                                if (index === -1) {
                                    attachments = mutate(attachments, {
                                        $push: [{ type: option.value, url: '' }]
                                    });
                                } else {
                                    attachments = mutate(attachments, {
                                        [index]: {
                                            $set: { type: option.value, url: attachment.url }
                                        }
                                    });
                                }
                                this.setState({ attachments });
                            }
                        }}
                        options={index > -1 ? TYPE_OPTIONS : NEW_TYPE_OPTIONS}
                    />
                </div>
                {index > -1 ? (
                    <>
                        <div className={styles.url}>
                            <TextInputElement
                                placeholder="URL"
                                name="url"
                                onChange={(value: string) => {
                                    attachments = mutate(attachments, {
                                        [index]: { $set: { type: attachment.type, url: value } }
                                    });
                                    this.setState({ attachments });
                                }}
                                entry={{ value: attachment.url }}
                                autocomplete={true}
                            />
                        </div>
                        <div className={styles.remove}>
                            <Pill
                                icon="fe-x"
                                text=" Remove"
                                large={true}
                                onClick={() => {
                                    this.handleAttachmentRemoved(index);
                                }}
                            />
                        </div>
                    </>
                ) : null}
            </div>
        );
    }

    private renderAttachments(): JSX.Element {
        const attachments = this.state.attachments.map(
            (attachment, index: number) =>
                attachment.uploaded
                    ? this.renderUpload(index, attachment)
                    : this.renderAttachment(index, attachment)
        );

        const emptyOption =
            this.state.attachments.length < MAX_ATTACHMENTS
                ? this.renderAttachment(-1, { url: '', type: '' })
                : null;
        return (
            <>
                <p>
                    Add up to {MAX_ATTACHMENTS} to each message. Each attachment can be a file you
                    upload or a dynamic URL using expressions and variables from your Flow.
                </p>
                {attachments}
                {emptyOption}
                <input
                    style={{
                        display: 'none'
                    }}
                    ref={ele => {
                        this.filePicker = ele;
                    }}
                    type="file"
                    onChange={e => this.handleUploadFile(e.target.files)}
                />
            </>
        );
    }

    public render(): JSX.Element {
        const typeConfig = this.props.typeConfig;

        const quickReplies: Tab = {
            name: 'Quick Replies',
            body: (
                <>
                    <p>
                        Quick Replies are made into buttons for supported channels. For example,
                        when asking a question, you might add a Quick Reply for "Yes" and one for
                        "No".
                    </p>
                    <TaggingElement
                        name="Replies"
                        placeholder="Quick Replies"
                        prompt="Enter a Quick Reply"
                        onChange={this.handleQuickRepliesUpdate}
                        onCheckValid={() => true}
                        entry={this.state.quickReplies}
                    />
                </>
            ),
            checked: this.state.quickReplies.value.length > 0
        };

        const attachments: Tab = {
            name: 'Attachments',
            body: this.renderAttachments(),
            checked: this.state.attachments.length > 0
        };

        const advanced: Tab = {
            name: 'Advanced',
            body: (
                <CheckboxElement
                    name="All Destinations"
                    title="All Destinations"
                    labelClassName={localStyles.checkbox}
                    checked={this.state.sendAll}
                    description="Send a message to all destinations known for this contact. If you aren't sure what this means, leave it unchecked."
                    onChange={this.handleSendAllUpdate}
                />
            ),
            checked: this.state.sendAll
        };

        return (
            <Dialog
                title={typeConfig.name}
                headerClass={typeConfig.type}
                buttons={this.getButtons()}
                tabs={[quickReplies, attachments, advanced]}
            >
                <TypeList
                    __className=""
                    initialType={typeConfig}
                    onChange={this.props.onTypeChange}
                />
                <TextInputElement
                    name="Message"
                    showLabel={false}
                    count={Count.SMS}
                    onChange={this.handleMessageUpdate}
                    entry={this.state.text}
                    autocomplete={true}
                    focus={true}
                    textarea={true}
                />
            </Dialog>
        );
    }
}
