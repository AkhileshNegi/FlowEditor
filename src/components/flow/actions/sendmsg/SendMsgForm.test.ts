import SendMsgForm from '~/components/flow/actions/sendmsg/SendMsgForm';
import { ActionFormProps } from '~/components/flow/props';
import { composeComponentTestUtils } from '~/testUtils';
import { createSendMsgAction, getActionFormProps } from '~/testUtils/assetCreators';

const { setup } = composeComponentTestUtils<ActionFormProps>(
    SendMsgForm,
    getActionFormProps(createSendMsgAction())
);

describe(SendMsgForm.name, () => {
    describe('render', () => {
        it('should render', () => {
            const { wrapper } = setup(true);
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('updates', () => {
        it('should save changes', () => {
            const { instance, props } = setup(true);

            instance.handleMessageUpdate('What is your favorite color?');
            instance.handleQuickRepliesUpdate(['red', 'green', 'blue']);
            instance.handleSendAllUpdate(true);

            expect(instance.state).toMatchSnapshot();

            instance.handleSave();
            expect(props.updateAction).toHaveBeenCalled();
            expect((props.updateAction as any).mock.calls[0]).toMatchSnapshot();
        });
    });

    describe('cancel', () => {
        it('should cancel without changes', () => {
            const { instance, props } = setup(true, {
                $merge: { onClose: jest.fn(), updateAction: jest.fn() }
            });
            instance.handleMessageUpdate("Don't save me bro");
            instance.getButtons().secondary.onClick();
            expect(props.onClose).toHaveBeenCalled();
            expect(props.updateAction).not.toHaveBeenCalled();
        });
    });
});
