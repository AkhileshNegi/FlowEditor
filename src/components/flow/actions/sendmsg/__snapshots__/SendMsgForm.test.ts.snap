// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SendMsgForm render should render 1`] = `
<Flipper
  back={
    <Dialog
      headerClass="send_msg"
      headerIcon="fe-cog"
      headerStyle="barber"
      subtitle="Advanced Settings"
      title="Send Message"
    >
      <p>
        Quick Replies are made into buttons for supported channels
      </p>
      <TaggingElement
        entry={
          Object {
            "value": Array [],
          }
        }
        name="Replies"
        onChange={[Function]}
        onCheckValid={[Function]}
        onValidPrompt={[Function]}
        placeholder="Quick Replies"
        prompt="Enter a Quick Reply"
      />
      <CheckboxElement
        checked={false}
        description="Send a message to all destinations known for this contact."
        labelClassName="checkbox"
        name="All Destinations"
        onChange={[Function]}
        title="All Destinations"
      />
    </Dialog>
  }
  front={
    <Dialog
      buttons={
        Object {
          "primary": Object {
            "name": "Ok",
            "onClick": [Function],
          },
          "secondary": Object {
            "name": "Cancel",
            "onClick": [Function],
          },
        }
      }
      headerClass="send_msg"
      title="Send Message"
    >
      <TypeList
        __className=""
        initialType={
          Object {
            "component": [Function],
            "description": "Send the contact a message",
            "form": undefined,
            "localization": [Function],
            "localizeableKeys": Array [
              "text",
              "quick_replies",
            ],
            "name": "Send Message",
            "type": "send_msg",
          }
        }
        onChange={[MockFunction]}
      />
      <Connect(TextInputElement)
        autocomplete={true}
        count="SMS"
        entry={
          Object {
            "value": "Hey!",
          }
        }
        focus={true}
        name="Message"
        onChange={[Function]}
        showLabel={false}
        textarea={true}
      />
    </Dialog>
  }
/>
`;

exports[`SendMsgForm updates should save changes 1`] = `
Object {
  "quickReplies": Object {
    "value": Array [
      "red",
      "green",
      "blue",
    ],
  },
  "sendAll": true,
  "text": Object {
    "value": "What is your favorite color?",
  },
  "valid": true,
}
`;

exports[`SendMsgForm updates should save changes 2`] = `
Array [
  Object {
    "all_urns": true,
    "quick_replies": Array [
      "red",
      "green",
      "blue",
    ],
    "text": "What is your favorite color?",
    "type": "send_msg",
    "uuid": "send_msg-0",
  },
]
`;
